ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java

```diff
@@ -0,0 +1,596 @@
+      throws Exception {
+    ArrayList<MockServer> mockServers = new ArrayList<>(mockServerLayout.getMockServers());
+    // set the status to various server level or partition level errors (not Blob_Deleted or Blob_Expired).
+    mockServers.get(0).setServerErrorForAllRequests(ServerErrorCode.Unknown_Error);
+    mockServers.get(1).setServerErrorForAllRequests(ServerErrorCode.Unknown_Error);
+    mockServers.get(2).setServerErrorForAllRequests(ServerErrorCode.IO_Error);
+    mockServers.get(3).setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+    mockServers.get(4).setServerErrorForAllRequests(ServerErrorCode.Data_Corrupt);
+    mockServers.get(5).setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+    mockServers.get(6).setServerErrorForAllRequests(ServerErrorCode.Blob_Not_Found);
+    mockServers.get(7).setServerErrorForAllRequests(ServerErrorCode.Disk_Unavailable);
+    mockServers.get(8).setServerErrorForAllRequests(ServerErrorCode.Unknown_Error);
+
+    // clear the hard error in one of the servers in the datacenter where the put happened.
+    for (int i = 0; i < mockServers.size(); i++) {
+      MockServer mockServer = mockServers.get(i);
+      if (mockServer.getDataCenter().equals(dcWherePutHappened)) {
+        mockServer.setServerErrorForAllRequests(ServerErrorCode.No_Error);
+        break;
+      }
+    }
+    getAndAssertSuccess();
+  }
+
+  // @todo: test the case where read is not called before chunks come in.
+  // @todo: test the case where read comes in immediately after callback, and chunks come in delayed.
+  // @todo: test the case where operation callback is called with exception.
+  // @todo: test the case where async write results in an exception. - read should be notified,
+  //        operation should get completed.
+  // @todo: test the case where a subsequent chunk fails at the router. the read should fail in that case.
+  // @todo: maybe a test for legacy blobs.
+  // @todo: possibly tests where intermediate chunks get expired/deleted.
+
+  /**
+   * Construct GetBlob operations with appropriate callbacks, then poll those operations until they complete,
+   * and ensure that the whole blob data is read out and the contents match.
+   */
+  private void getAndAssertSuccess()
+      throws Exception {
+    final CountDownLatch readCompleteLatch = new CountDownLatch(1);
+    final AtomicReference<Exception> readCompleteException = new AtomicReference<>(null);
+    final AtomicLong readCompleteResult = new AtomicLong(0);
+    final AtomicReference<Exception> operationException = new AtomicReference<>(null);
+
+    Callback<ReadableStreamChannel> callback = new Callback<ReadableStreamChannel>() {
+      @Override
+      public void onCompletion(final ReadableStreamChannel result, final Exception exception) {
+        if (exception != null) {
+          operationException.set(exception);
+          readCompleteLatch.countDown();
+        } else {
+          Utils.newThread(new Runnable() {
```

Comment:
Not sure if I'm entirely correct but I suspect that if is possible to get rid of this thread, then you would avoid waiting on latches because everything would be done synchronously.